server/src/fstar_handlers.ts

```diff
@@ -0,0 +1,403 @@
+import {
+	Diagnostic,
+	DiagnosticRelatedInformation,
+	DiagnosticSeverity,
+	Range
+} from 'vscode-languageserver/node';
+
+import {
+	TextDocument
+} from 'vscode-languageserver-textdocument';
+
+import * as crypto from 'crypto';
+
+import { fstarVSCodeAssistantSettings } from './settings';
+import { Server } from './server';
+import { ClientConnection, StatusOkMessage, ok_kind } from './client_connection';
+import { mkPosition, fstarRangeAsRange, qualifyFilename } from './utils';
+
+// Cyclic dependency to support mocking out functions within this module when
+// testing other functions within the module. Suggested as a solution in this
+// post: https://stackoverflow.com/questions/51269431/jest-mock-inner-function
+import * as fstar_handlers from './fstar_handlers';
+
+///////////////////////////////////////////////////////////////////////////////////
+// Handling responses from the F* IDE protocol
+///////////////////////////////////////////////////////////////////////////////////
+
+// All messages from F* are expected to be valid JSON objects.
+//
+// TODO(klinvill): this should likely be refactored into `fstar_messages.ts` and
+// should check the structure of a message, not just that it's valid JSON. A
+// better method could return either the appropriate message object, or an error
+// otherwise, so that the parsing could be moved out of these handlers and into
+// the same file as the message definitions.
+function is_valid_fstar_message(entry: string): boolean {
+	try {
+		JSON.parse(entry);
+		return true;
+	}
+	catch (err) {
+		return false;
+	}
+}
+
+// Event handler for stdout on fstar_ide. Created as a closure to keep the
+// buffer scoped only to this function. The factory function exists to make
+// unit-testing easier (creating a new function is like resetting the closure
+// state).
+export function handleFStarResponseForDocumentFactory(configurationSettings: fstarVSCodeAssistantSettings, server: Server, connection: ClientConnection): ((textDocument: TextDocument, data: string, lax: boolean) => void) {
+	// Stateful buffer to store partial messages. Messages appear to be fragmented
+	// into 8192 byte chunks if they exceed this size.
+	let buffer = "";
+
+	const handleOneResponseForDocument = fstar_handlers.handleOneResponseForDocumentFactory(configurationSettings, server, connection);
+	return function (textDocument: TextDocument, data: string, lax: boolean) {
+		if (configurationSettings.debug) {
+			console.log("<<< " + (lax ? "lax" : "") + "uri:<" + textDocument.uri + ">:" + data);
+		}
+		const lines = data.toString().split('\n');
+
+		const valid_lines: string[] = [];
+		for (const line of lines) {
+			if (is_valid_fstar_message(line)) {
+				// We assume that fragmented messages will always be delivered
+				// sequentially. Because of this, receiving a non-fragmented
```

Comment:
Thanks for the feedback. I still think it's valuable to have a comment here documenting the assumptions that messages will always be delivered in-order, just in case the behavior changes in the future, like if the F* IDE is reworked to be multi-threaded. I tried to incorporate some of this explanation into the comment. Do you think it's clearer now, or are there other changes you'd like to see made?

Reply:
Yes, it's clearer now.